exercises/bank-account/bank_account_test.py

```diff
@@ -0,0 +1,125 @@
+import unittest
+import threading
+
+from bank_account import BankAccount
+
+
+class BankAccountTests(unittest.TestCase):
+
+    def test_newly_opened_account_has_zero_balance(self):
+        self.account = BankAccount()
+        self.account.open()
+
+        self.assertEqual(self.account.getBalance(), 0)
+
+    def test_get_balance_0(self):
+        self.account = BankAccount()
+        self.account.open()
+
+        self.assertEqual(self.account.getBalance(), 0)
+
+    def test_get_balance_with_amount(self):
+        self.account = BankAccount()
+        self.account.open()
+        self.account.deposit(100)
+
+        self.assertEqual(self.account.getBalance(), 100)
+
+    def test_deposit_into_account(self):
+        self.account = BankAccount()
+        self.account.open()
+        self.account.deposit(100)
+
+        self.assertEqual(self.account.getBalance(), 100)
+
+    def test_two_deposits_in_a_row(self):
+        self.account = BankAccount()
+        self.account.open()
+        self.account.deposit(100)
+        self.account.deposit(50)
+
+        self.assertEqual(self.account.getBalance(), 150)
+
+    def test_can_withdraw(self):
+        self.account = BankAccount()
+        self.account.open()
+        self.account.deposit(100)
+        self.account.withdraw(50)
+
+        self.assertEqual(self.account.getBalance(), 50)
+
+    def test_checking_balance_of_closed_account_throws_error(self):
+        self.account = BankAccount()
+        self.account.open()
+        self.account.close()
+
+        with self.assertRaises(ValueError):
+            self.account.getBalance()
+
+    def test_deposit_into_closed_account(self):
+        self.account = BankAccount()
+        self.account.open()
+        self.account.close()
+
+        with self.assertRaises(ValueError):
+            self.account.deposit(50)
+
+    def test_withdraw_from_closed_account(self):
+        self.account = BankAccount()
+        self.account.open()
+        self.account.close()
+
+        with self.assertRaises(ValueError):
+            self.account.withdraw(50)
+
+    def test_cannot_withdraw_more_than_deposited(self):
+        self.account = BankAccount()
+        self.account.open()
+        self.account.deposit(25)
+
+        with self.assertRaises(ValueError):
+            self.account.withdraw(50)
+
+    def test_cannot_withdraw_negative(self):
+        self.account = BankAccount()
+        self.account.open()
+        self.account.deposit(100)
+
+        with self.assertRaises(ValueError):
+            self.account.withdraw(-50)
+
+    def test_cannot_deposit_negative(self):
+        self.account = BankAccount()
+        self.account.open()
+
+        with self.assertRaises(ValueError):
+            self.account.deposit(-50)
+
+    def test_can_handle_concurrent_transactions(self):
```

Comment:
@aes421 (cc: @cmccandless, @m-a-ge)

I'm not sure that this test is truly testing for concurrency - depositing and then immediately withdrawing inherently safe values within one function is never going to break as far as I can see. Further, I can remove all instances of `self.lock` from the example code and the tests still pass. Part of the issue is probably that the Global Interpreter Lock (GIL) already locks the `BankAccount` instance when a thread accesses it, and releases it once the thread ends. I'm not sure of the best way to get around this.

Currently, Python seems to be running these threads in order, depositing `10` and withdrawing `1`, then moving on to the next thread. We can see this by storing a list of transaction values and using `i+1` as a second argument for the `increment_and_decrement` function, depositing and withdrawing this value, and then printing the transaction list after all threads are done.

One possible solution to get disorder seems to be to add `time.sleep(random.random())` before deposits or withdrawals, but I think the deposits and withdrawal should be decoupled and not part of the same function call - just joining a list of threads for each should be sufficient if there are random delays.

Concurrency was discussed a little in #1106. If we want to keep concurrency in this exercise (I think the exercise would be ok without it, but only as an early exercise about classes), I think that some consideration needs to be made for how to make the tests only pass if it is safe concurrent implementation.

Ultimately, for concurrency to make sense here, we need to be able to make it so that the concurrency matters in some way.

Reply:
@N-Parsons (cc: @aes421, @m-a-ge)

>I think the exercise would be ok without it

I disagree. It's a core part of the exercise. From the [problem description](https://github.com/exercism/problem-specifications/blob/4d2489f8502614e49a8f041f23dd85600d65658f/exercises/bank-account/description.md):
>Create an account that can be accessed from **multiple threads/processes** (terminology depends on your programming language).

If we marked another exercise as foregone simply because of the challenges of testing parallelism in Python, then I think we might need to consider foregoing this exercise as well. Either that, or we need to come up with another solution and apply it both to this exercise and the foregone `parallel-letter-frequency`.

Reply:
If we want to be pedantic with this exercise, we could do something similar to NumPy and write an extension in C that runs outside of the GIL. However we can only safely do that if we use no Python C API function used.

I also want to quickly put a note here because I don't think concurrency and parallelism have been discussed much in this repo. There is a difference between concurrency and parallelism.

"In programming, concurrency is the composition of independently executing processes, while parallelism is the simultaneous execution of (possibly related) computations." - from golang.org

https://vimeo.com/49718712 (Video of Rob Pike talking about the difference at Heroku)

Reply:
We also might be able to use Pools:

https://docs.python.org/2/library/multiprocessing.html